### PR TITLE
Issue #110: Split Blocking IP setting.

### LINKED ISSUE #110: Split ip blocking into two settings one which can be forced
https://github.com/catalyst/moodle-tool_heartbeat/blob/master/settings.php#L46

### PR DESCRIPTION
**Pull request by jaypha**
Resolves #110 

Add lib class for general functions
Add lib::get_allowed_ips() function
Modify iplock.php to use lib::get_allowed_ips()
Add unit test for lib class.
Version bump

### CODE REVIEW

File: tests/lib_test.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * Test class for tool_heartbeat\lib
+ *
+ * @package   tool_heartbeat
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+
```
Comment (keevan) [inline tests/lib_test.php:28]: Can you please remove the unnecessary whitespace here.

File: classes/lib.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_heartbeat;
+
+/**
+ * General fucntion for use with heartbeat.
+ *
```
Comment (keevan) [inline classes/lib.php:21]: Can you please remove the unnecessary whitespace here.

File: lang/en/tool_heartbeat.php [hunk 3]
```diff
@@ -43,12 +43,16 @@
 $string['configauthmissing'] = 'Configured auth methods are not currently enabled.';
 $string['setinitialauthstate'] = 'Initial auth state for heartbeat auth check set.';
 $string['authcorrect'] = 'Auth methods correctly configured.';
+$string['builtinallowediplist'] = 'Builtin IP Blocking Configuration';
+$string['builtinallowediplist_desc'] = 'A second allowed IP list which makes it easier to have some IPs forced in config.php and others editable in the UI';
```
Comment (keevan) [inline lang/en/tool_heartbeat.php:47]:
```suggestion
$string['builtinallowediplist_desc'] = 'This allowed IP list would allow some IPs to be editable in the UI in addition to those forced in config.php';
```

File: tests/lib_test.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+class lib_test extends advanced_testcase {
+    /**
+     * Test lib::test_get_allowed_ips
+     * @throws coding_exception
+     * @throws dml_exception
```
[the remark this answers is not in the document]
Comment (jaypha) [reply tests/lib_test.php]: Done

File: tests/lib_test.php (hunk 1 above)
Comment (jaypha) [reply tests/lib_test.php]: Done

[the remark this answers is not in the document]
Comment (jaypha) [reply tests/lib_test.php]: Done

File: classes/lib.php (hunk 2 above)
Comment (jaypha) [reply classes/lib.php:21]: Done

File: classes/lib.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+class lib {
+    /**
+     * Return the list of allowed IPs, by combining the UI editable IP list with the
+     * config defined IP list.
+     *
+     * @return string
+     * @throws \coding_exception
+     * @throws \dml_exception
```
[the remark this answers is not in the document]
Comment (jaypha) [reply classes/lib.php]: Done

File: lang/en/tool_heartbeat.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (jaypha) [reply lang/en/tool_heartbeat.php]: Done

File: classes/lib.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+     */
+    public static function get_allowed_ips(): string {
+        return trim(
+            get_config('tool_heartbeat', 'allowedips') .
+            "\n" .
```
[the remark this answers is not in the document]
Comment (jaypha) [reply classes/lib.php]: Done

File: settings.php [hunk 7]
```diff
@@ -56,7 +56,7 @@
             $type = 'notifyerror';
         };
         $description .= $OUTPUT->notification(get_string($message, 'tool_heartbeat', ['ip' => getremoteaddr()]), $type);
-
+        $description .= '<p>'.get_string('ips_combine', 'tool_heartbeat').'</p>';
```
[the remark this answers is not in the document]
Comment (jaypha) [reply settings.php]: Done

File: lang/en/tool_heartbeat.php (hunk 3 above)
Comment (jaypha) [reply lang/en/tool_heartbeat.php]: Done

File: tests/lib_test.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * Test class for tool_heartbeat\lib
+ *
+ * @package   tool_heartbeat
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
```
Comment (keevan) [inline tests/lib_test.php:27]: Can you please remove the unnecessary whitespace here.

Comment (keevan) [reply tests/lib_test.php:27]: After this, happy for it to be merged.
